File: HLPProject/HLPProject/Conditions.fs
// This module details the instruction condition codes.

namespace Common
module Conditions =
    
    open State

    // Condition functions
    /// Always condition.
    let checkAL state =
        true

    /// Equal condition.
    let checkEQ state =
        readZFlag state

    /// Not equal condition.
    let checkNE state =
        not (readZFlag state)

    /// Carry set condition.
    let checkCS state =
        readCFlag state

    /// Carry clear condition.
    let checkCC state =
        not (readCFlag state)

    /// Minus condition.
    let checkMI state =
        readNFlag state

    /// Plus condition.
    let checkPL state =
        not (readNFlag state)

    /// Overflow set condition
    let checkVS state =
        readVFlag state

    /// Overflow clear condition
    let checkVC state =
        not (readVFlag state)

    /// HI condition.
    let checkHI state =
        (readCFlag state) && (not (readZFlag state))

    /// LS condition.
    let checkLS state =
        (not (readCFlag state)) || (readZFlag state)

    /// Greater than or equal condition.
    let checkGE state =
        (readNFlag state) = (readVFlag state)

    /// Less than condition.
    let checkLT state =
        (readNFlag state) <> (readVFlag state)

    /// Greater than condition.
    let checkGT state =
        (not (readZFlag state)) && ((readNFlag state) = (readVFlag state))

    /// Less than or equal condition.
    let checkLE state =
        (readZFlag state) && ((readNFlag state) <> (readVFlag state))

File: HLPProject/HLPProject/hlp/Conditions.js
import { readVFlag, readNFlag, readCFlag, readZFlag } from "./State";
export function checkAL(state) {
  return true;
}
export function checkEQ(state) {
  return readZFlag(state);
}
export function checkNE(state) {
  return !readZFlag(state);
}
export function checkCS(state) {
  return readCFlag(state);
}
export function checkCC(state) {
  return !readCFlag(state);
}
export function checkMI(state) {
  return readNFlag(state);
}
export function checkPL(state) {
  return !readNFlag(state);
}
export function checkVS(state) {
  return readVFlag(state);
}
export function checkVC(state) {
  return !readVFlag(state);
}
export function checkHI(state) {
  if (readCFlag(state)) {
    return !readZFlag(state);
  } else {
    return false;
  }
}
export function checkLS(state) {
  if (!readCFlag(state)) {
    return true;
  } else {
    return readZFlag(state);
  }
}
export function checkGE(state) {
  return readNFlag(state) === readVFlag(state);
}
export function checkLT(state) {
  return readNFlag(state) !== readVFlag(state);
}
export function checkGT(state) {
  if (!readZFlag(state)) {
    return readNFlag(state) === readVFlag(state);
  } else {
    return false;
  }
}
export function checkLE(state) {
  if (readZFlag(state)) {
    return readNFlag(state) !== readVFlag(state);
  } else {
    return false;
  }
}